File: apps/jetbrains/src/main/kotlin/icu/seismic/plugin/Detector.kt
package icu.seismic.plugin

import java.io.File
import com.intellij.openapi.application.ApplicationNamesInfo
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import java.net.InetAddress
import java.util.TimeZone
import java.util.concurrent.TimeUnit

object Detector {
    fun detectProject(project: Project, useGitRoot: Boolean = false): String {
        if (useGitRoot) {
            detectGitRoot(project)?.let { return File(it).name }
        }
        return project.name
    }

    fun detectGitRoot(project: Project): String? {
        val basePath = project.basePath ?: return null
        return runGit(basePath, listOf("rev-parse", "--show-toplevel"))
    }

    /** Shells out to git directly — works regardless of whether the IDE's
     * own VCS integration is configured for the project. */
    fun detectBranch(project: Project): String? {
        val basePath = project.basePath ?: return null
        return try {
            runGit(basePath, listOf("rev-parse", "--abbrev-ref", "HEAD"))
        } catch (_: Exception) {
            null // git not installed, not a repo, etc — just skip branch info
        }
    }

    fun detectProjectMetadata(project: Project, useGitRoot: Boolean): ProjectSyncPayload? {
        val basePath = project.basePath ?: return null
        val projectPath = if (useGitRoot) detectGitRoot(project) ?: basePath else basePath
        val commits = detectCommits(projectPath)

        return ProjectSyncPayload(
            project = File(projectPath).name.ifEmpty { project.name },
            repoUrl = detectRepoUrl(projectPath),
            websiteUrl = detectWebsiteUrl(projectPath),
            lastCommitAt = commits.firstOrNull()?.committedAt,
            commits = commits,
        )
    }

    private fun detectRepoUrl(projectPath: String): String? {
        val url = runGit(projectPath, listOf("config", "--get", "remote.origin.url")) ?: return null
        val sshMatch = Regex("^git@([^:]+):(.+)$").find(url)
        if (sshMatch != null) {
            val host = sshMatch.groupValues[1]
            val repo = sshMatch.groupValues[2].removeSuffix(".git")
            return "https://$host/$repo"
        }
        return url.removeSuffix(".git")
    }

    private fun detectWebsiteUrl(projectPath: String): String? {
        val packageJson = File(projectPath, "package.json")
        if (!packageJson.isFile) return null

        val text = runCatching { packageJson.readText() }.getOrNull() ?: return null
        val homepage = Regex("\"homepage\"\\s*:\\s*\"([^\"]+)\"").find(text)?.groupValues?.get(1)
        val website = Regex("\"website\"\\s*:\\s*\"([^\"]+)\"").find(text)?.groupValues?.get(1)
        return homepage ?: website
    }

    private fun detectCommits(projectPath: String, limit: Int = 20): List<ProjectCommitPayload> {
        val output = runGit(projectPath, listOf("log", "-$limit", "--format=%H%x1f%ct%x1f%an%x1f%ae%x1f%s"))
            ?: return emptyList()

        return output.lineSequence()
            .filter { it.isNotBlank() }
            .mapNotNull { line ->
                val parts = line.split('\u001f')
                if (parts.isEmpty() || parts[0].isBlank()) return@mapNotNull null
                ProjectCommitPayload(
                    hash = parts[0],
                    committedAt = parts.getOrNull(1)?.toLongOrNull()?.times(1000),
                    authorName = parts.getOrNull(2),
                    authorEmail = parts.getOrNull(3),
                    message = parts.getOrNull(4),
                )
            }
            .toList()
    }

    private fun runGit(projectPath: String, args: List<String>): String? {
        return try {
            val process = ProcessBuilder(listOf("git", "-C", projectPath) + args)
                .redirectErrorStream(true)
                .start()
            if (!process.waitFor(2, TimeUnit.SECONDS)) {
                process.destroyForcibly()
                return null
            }
            if (process.exitValue() != 0) return null
            process.inputStream.bufferedReader().readText().trim().takeIf { it.isNotEmpty() }
        } catch (_: Exception) {
            null
        }
    }

    fun detectOS(): String {
        val name = System.getProperty("os.name").lowercase()
        return when {
            name.contains("win") -> "windows"
            name.contains("mac") -> "darwin"
            else -> "linux"
        }
    }

    fun detectMachine(): String = try {
        InetAddress.getLocalHost().hostName
    } catch (_: Exception) {
        "unknown"
    }

    fun detectTimezone(): String = TimeZone.getDefault().id

    /** Returns the IDE's own product name — e.g. "WebStorm", "GoLand",
     * "IntelliJ IDEA", "RustRover" — same idea as VS Code's editor: 'vscode',
     * but reflects whichever specific IDE is actually running. */
    fun detectEditorName(): String = ApplicationNamesInfo.getInstance().fullProductName

    fun shouldTrack(file: VirtualFile?): Boolean {
        if (file == null) return false
        if (!file.isInLocalFileSystem) return false
        if (!file.exists()) return false
        return true
    }

    fun languageId(file: VirtualFile): String = file.fileType.name.lowercase()
}

File: apps/nvim/lua/seismic/detector.lua
local M = {}

local function systemlist(args)
	local result = vim.fn.systemlist(args)
	if vim.v.shell_error ~= 0 or #result == 0 then
		return nil
	end
	return result
end

function M.detect_git_root()
	local result = systemlist({ "git", "rev-parse", "--show-toplevel" })
	return result and result[1] or nil
end

function M.detect_project(use_git_root)
	if use_git_root then
		local git_root = M.detect_git_root()
		if git_root then
			return vim.fn.fnamemodify(git_root, ":t")
		end
	end

	local cwd = vim.fn.getcwd()
	return vim.fn.fnamemodify(cwd, ":t")
end

function M.detect_branch()
	local result = systemlist({ "git", "rev-parse", "--abbrev-ref", "HEAD" })
	return result and result[1] or nil
end

function M.detect_repo_url(project_path)
	local result = systemlist({ "git", "-C", project_path, "config", "--get", "remote.origin.url" })
	local url = result and result[1] or nil
	if not url or url == "" then
		return nil
	end
	local host, repo = url:match("^git@([^:]+):(.+)$")
	if host and repo then
		return "https://" .. host .. "/" .. repo:gsub("%.git$", "")
	end
	return url:gsub("%.git$", "")
end

function M.detect_website_url(project_path)
	local package_json = project_path .. "/package.json"
	if vim.fn.filereadable(package_json) ~= 1 then
		return nil
	end

	local ok, body = pcall(vim.json.decode, table.concat(vim.fn.readfile(package_json), "\n"))
	if not ok or type(body) ~= "table" then
		return nil
	end
	return body.homepage or body.website
end

function M.detect_commits(project_path, limit)
	limit = limit or 20
	local result = systemlist({
		"git",
		"-C",
		project_path,
		"log",
		"-" .. tostring(limit),
		"--format=%H%x1f%ct%x1f%an%x1f%ae%x1f%s",
	})
	if not result then
		return {}
	end

	local commits = {}
	for _, line in ipairs(result) do
		local parts = vim.split(line, "\31", { plain = true })
		table.insert(commits, {
			hash = parts[1],
			committedAt = parts[2] and tonumber(parts[2]) and tonumber(parts[2]) * 1000 or nil,
			authorName = parts[3],
			authorEmail = parts[4],
			message = parts[5],
		})
	end
	return commits
end

function M.detect_project_metadata(use_git_root)
	local cwd = vim.fn.getcwd()
	local project_path = use_git_root and M.detect_git_root() or nil
	project_path = project_path or cwd
	local commits = M.detect_commits(project_path)

	return {
		project = vim.fn.fnamemodify(project_path, ":t"),
		repoUrl = M.detect_repo_url(project_path),
		websiteUrl = M.detect_website_url(project_path),
		lastCommitAt = commits[1] and commits[1].committedAt or nil,
		commits = commits,
	}
end

function M.detect_os()
	local sysname = vim.uv.os_uname().sysname
	if sysname == "Linux" then
		return "linux"
	end
	if sysname == "Darwin" then
		return "darwin"
	end
	if sysname:match("Windows") then
		return "windows"
	end
	return sysname:lower()
end

function M.detect_machine()
	return vim.uv.os_gethostname()
end

function M.detect_timezone()
	local tz = os.date("%Z")
	return tz or ""
end

function M.language_id(bufnr)
	local ft = vim.bo[bufnr].filetype
	return ft ~= "" and ft or "text"
end

function M.should_track(bufnr)
	if vim.bo[bufnr].buftype ~= "" then
		return false
	end -- skip terminal, quickfix, etc
	local name = vim.api.nvim_buf_get_name(bufnr)
	return name ~= ""
end

return M
